File: src/main/java/no/novari/flyt/configuration/validation/instancereference/InstanceValueKeyExtractionService.java
package no.novari.flyt.configuration.validation.instancereference;

import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Collections;
import java.util.regex.MatchResult;
import java.util.regex.Pattern;

@Service
public class InstanceValueKeyExtractionService {

    private static final Pattern ifReferenceExtractionPattern = Pattern.compile("\\$if\\{[^}]+}");

    public Collection<String> extractIfReferenceKeys(String value) {
        return value == null
                ? Collections.emptyList()
                : ifReferenceExtractionPattern.matcher(value)
                .results()
                .map(MatchResult::group)
                .map(ifReference -> ifReference.replace("$if{", "").replace("}", ""))
                .toList();
    }
}

File: src/main/kotlin/no/novari/flyt/configuration/validation/instancereference/InstanceValueKeyExtractionService.kt
package no.novari.flyt.configuration.validation.instancereference

import org.springframework.stereotype.Service
import java.util.regex.MatchResult
import java.util.regex.Pattern

@Service
class InstanceValueKeyExtractionService {
    private val referenceExtractionPattern = Pattern.compile("${Pattern.quote(IF_REFERENCE_PREFIX)}[^}]+}")

    fun extractIfReferenceKeys(value: String?): Collection<String> =
        value
            ?.let { content ->
                referenceExtractionPattern
                    .matcher(content)
                    .results()
                    .map(MatchResult::group)
                    .map { ifReference -> ifReference.removePrefix(IF_REFERENCE_PREFIX).removeSuffix("}") }
                    .toList()
            }
            ?: emptyList()

    private companion object {
        const val IF_REFERENCE_PREFIX = "\$if{"
    }
}
